hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/repository/TokenBalanceRepository.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.mirror.web3.repository;
+
+import com.hedera.mirror.common.domain.balance.TokenBalance;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.CrudRepository;
+
+public interface TokenBalanceRepository extends CrudRepository<TokenBalance, TokenBalance.Id> {
+
+    /**
+     * Retrieves the token balance of a specific token for a specific account at the last consensus timestamp
+     * before the provided block timestamp.
+     *
+     * @param tokenId         the ID of the token whose balance is to be retrieved.
+     * @param accountId       the ID of the account whose balance is to be retrieved.
+     * @param blockTimestamp  the block timestamp used as the upper limit to retrieve the token balance.
+     *                        The method will retrieve the last token balance before this timestamp.
+     * @return an Optional containing the balance if found, or an empty Optional if no matching balance
+     *         entry is found before the given block timestamp.
+     */
+    @Query(
+            value =
+                    """
+                select * from token_balance
+                where
+                token_id = ?1 and
+                account_id = ?2 and
+                consensus_timestamp < ?3
+                order by consensus_timestamp desc
+                limit 1
```

Comment:
nit, this is slightly better indentation
```suggestion
                select * from token_balance
                where
                  token_id = ?1 and
                  account_id = ?2 and
                  consensus_timestamp < ?3
                order by consensus_timestamp desc
                limit 1
```

Reply:
Comment addressed.